exporter/kafkaexporter/pdata_marshaler.go

```diff
@@ -82,17 +85,30 @@ type pdataTracesMarshaler struct {
 	encoding  string
 }
 
-func (p pdataTracesMarshaler) Marshal(td pdata.Traces, topic string) ([]*sarama.ProducerMessage, error) {
-	bts, err := p.marshaler.MarshalTraces(td)
-	if err != nil {
-		return nil, err
-	}
-	return []*sarama.ProducerMessage{
-		{
+func (p pdataTracesMarshaler) Marshal(td pdata.Traces, topic string) (messages []*sarama.ProducerMessage, errs error) {
+	for _, batch := range batchpersignal.SplitTraces(td) {
+		bts, err := p.marshaler.MarshalTraces(td)
+		if err != nil {
+			// continue to process batches that can be serialized
+			errs = multierr.Append(errs, err)
+			continue
+		}
+
+		var key sarama.Encoder
+		if batch.ResourceSpans().Len() > 0 &&
+			batch.ResourceSpans().At(0).InstrumentationLibrarySpans().Len() > 0 &&
+			batch.ResourceSpans().At(0).InstrumentationLibrarySpans().At(0).Spans().Len() > 0 {
+			traceID := batch.ResourceSpans().At(0).InstrumentationLibrarySpans().At(0).Spans().At(0).TraceID().Bytes()
+			key = sarama.ByteEncoder(traceID[:])
+		}
+
+		messages = append(messages, &sarama.ProducerMessage{
 			Topic: topic,
+			Key:   key,
```

Comment:
Now, this is a considerable change and a change of behaviour, I think in order to maintain backwards behaviour this should be user configurable if they want the data key-ed or not.

Basically instead of a randomised distribution, there might be hot partitions being written to which could end up as a bad time.

Reply:
I think this is my only thing I'd like to be configurable in this change @rcrowe :) 

More because murdering someone's Kafka cluster isn't a great time.

Reply:
Yeah, I kinda went with "well we do it with Jaeger", but v.good point this could regress people's usage.
I kinda wanted to skirt configuration, for now, to keep things simple; blocker?

Reply:
I'll go play around with how best to pass in the config. 

Marshallers are created before the factory has parsed any config
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b3526161a272dbb51453c01bce6279e75f92895b/exporter/kafkaexporter/factory.go#L66

Reply:
If you modify the type `pdataLogsMarshaler` to include a flag that is basically `"keyed"` and in future, we can use our ideas of allowing configurable key by method and that should be enough :) 

For the time being it can just be:

```
func (p pdataTracesMarshaler) Marshal(td pdata.Traces, topic string) (messages []*sarama.ProducerMessage, errs error) {
  if !p.Keyed {
    // Do old method of publishing
  }
  // Perform key-ing operation
}